r2/r2/lib/media.py

```diff
@@ -205,6 +452,83 @@ def optimize_jpeg(filename):
     with open(os.path.devnull, 'w') as devnull:
         subprocess.check_call(("/usr/bin/jpegoptim", filename), stdout=devnull)
 
+class _SVGSoup(BeautifulSoup.BeautifulStoneSoup):
+    # Fixes unexpected conversion from `<g><g></g></g>` into `<g></g><g></g>`
+    nests = ('defs', 'g', 'marker', 'mask', 'pattern', 'svg', 'symbol', 'use')
+    NESTABLE_TAGS = BeautifulSoup.buildTagMap([], nests)
+
+def _sanitize_svg(image):
+    elements = [x.lower() for x in _SAFE_SVG_ELEMENTS]
+    attributes = [x.lower() for x in _SAFE_SVG_ATTRIBUTES]
+
+    # BS3 gets confused with some self-closing tags. We can safely 
+    # convert them with regex and afterward, we strip out unwanted <!> 
+    # and <?> tags.
+    #
+    # Keep in mind that `[^>]` goes in place of `.`, because it matches
+    # newline chars as well. One could also flag re.DOTALL, but we need
+    # to also not match `>`s.
+    image = re.sub(r'<([^>\s]+)\s*([^>]*)/>', r'<\1 \2></\1>',
+                   re.sub(r'<![^>]*>|<\?(?!xml)[^>]*>', r'', image))
```

Comment:
I chose to use BS3 because it was already imported and was able to do _most_ of the job I wanted. I would be happy to use `defusedxml.lxml`, but I wasn't really sure what the policy was on adding new packages -- nor how exactly to do it.